core/src/main/scala/org/apache/spark/util/DependencyUtils.scala

```diff
@@ -15,22 +15,122 @@
  * limitations under the License.
  */
 
-package org.apache.spark.deploy
+package org.apache.spark.util
 
 import java.io.File
-import java.net.URI
+import java.net.{URI, URISyntaxException}
 
 import org.apache.commons.lang3.StringUtils
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.{FileSystem, Path}
 
 import org.apache.spark.{SecurityManager, SparkConf, SparkException}
+import org.apache.spark.deploy.SparkSubmitUtils
 import org.apache.spark.internal.Logging
-import org.apache.spark.util.{MutableURLClassLoader, Utils}
 
-private[deploy] object DependencyUtils extends Logging {
+private[spark] object DependencyUtils extends Logging {
+
+  def getIvyProperties(): Seq[String] = {
+    Seq(
+      "spark.jars.excludes",
+      "spark.jars.packages",
+      "spark.jars.repositories",
+      "spark.jars.ivy",
+      "spark.jars.ivySettings"
+    ).map(sys.props.get(_).orNull)
+  }
+
+  /**
+   * Parse excluded list in ivy URL. When download ivy URL jar, Spark won't download transitive jar
+   * in excluded list.
+   *
+   * @param queryString Ivy URI query part string.
+   * @return Exclude list which contains grape parameters of exclude.
+   *         Example: Input:  exclude=org.mortbay.jetty:jetty,org.eclipse.jetty:jetty-http
+   *         Output:  [org.mortbay.jetty:jetty, org.eclipse.jetty:jetty-http]
+   */
+  private def parseExcludeList(excludes: Array[String]): String = {
+    excludes.flatMap { excludeString =>
+      val excludes: Array[String] = excludeString.split(",")
+      if (excludes.exists(_.split(":").length != 2)) {
+        throw new URISyntaxException(excludeString,
+          "Invalid exclude string: expected 'org:module,org:module,..', found " + excludeString)
+      }
+      excludes
+    }.mkString(":")
+  }
+
+  /**
+   * Parse transitive parameter in ivy URL, default value is false.
+   *
+   * @param queryString Ivy URI query part string.
+   * @return Exclude list which contains grape parameters of transitive.
+   *         Example: Input:  exclude=org.mortbay.jetty:jetty&transitive=true
+   *         Output:  true
+   */
+  private def parseTransitive(transitives: Array[String]): Boolean = {
+    if (transitives.isEmpty) {
+      false
+    } else {
+      if (transitives.length > 1) {
+        logWarning("It's best to specify `transitive` parameter in ivy URL query only once." +
+          " If there are multiple `transitive` parameter, we will select the last one")
+      }
+      transitives.last.toBoolean
```

Comment:
What if `transitive=invalidStr` in hive? Could you add tests?

Reply:
Added in `test("SPARK-33084: Add jar support ivy url")`
```
    // Invalid transitive value, will use default value `false`
    sc.addJar("ivy://org.scala-js:scalajs-test-interface_2.12:1.2.0?transitive=foo")
    assert(!sc.listJars().exists(_.contains("scalajs-library_2.12")))
    assert(sc.listJars().exists(_.contains("scalajs-test-interface_2.12")))

```

Reply:
> What if `transitive=invalidStr` in hive? Could you add tests?

I change the logical, make it like hive. avoid convert error in `toBoolean`

Reply:
The current behaviour is the same with hive?